scanpipe/cyclonedx/__init__.py

```diff
@@ -0,0 +1,181 @@
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+#
+# Data Generated with ScanCode.io is provided on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, either express or implied. No content created from
+# ScanCode.io should be considered or used as legal advice. Consult an Attorney
+# for any legal advice.
+#
+# ScanCode.io is a free software code scanning tool from nexB Inc. and others.
+# Visit https://github.com/nexB/scancode.io for support and download.
+
+import json
+import pathlib
+
+import jsonschema
+from hoppr_cyclonedx_models.cyclonedx_1_4 import Component
+from hoppr_cyclonedx_models.cyclonedx_1_4 import (
+    CyclonedxSoftwareBillOfMaterialsStandard as Bom_1_4,
+)
+
+CYCLONEDX_SPEC_VERSION = "1.4"
+CYCLONEDX_JSON_SCHEMA_LOCATION = "bom-1.4.schema.json"
+CYCLONEDX_JSON_SCHEMA_PATH = (
+    pathlib.Path(__file__).parent / CYCLONEDX_JSON_SCHEMA_LOCATION
+)
+CYCLONEDX_JSON_SCHEMA_URL = (
+    "https://raw.githubusercontent.com/"
+    "CycloneDX/specification/master/schema/bom-1.4.schema.json"
+)
+
+
+def get_bom(cyclonedx_document: dict):
+    """
+    Return CycloneDx BOM object.
+    """
+    return Bom_1_4(**cyclonedx_document)
+
+
+def get_components(bom: Bom_1_4):
+    """
+    Return list of components from CycloneDx BOM.
+    """
+    return recursive_component_collector(bom.components, [])
+
+
+def bom_attributes_to_dict(cyclonedx_attributes):
+    """
+    Return list dict from a list of CycloneDx attributes.
+    """
+    if not cyclonedx_attributes:
+        return {}
+
+    return [
+        json.loads(attribute.json(exclude_unset=True, by_alias=True))
+        for attribute in cyclonedx_attributes
+    ]
+
+
+def recursive_component_collector(root_component_list, collected):
+    """
+    Return list of components including the nested components.
+    """
+    if not root_component_list:
+        return
+
+    for component in root_component_list:
+        extra_data = (
+            bom_attributes_to_dict(component.components)
+            if component.components is not None
+            else {}
+        )
+        collected.append({"cdx_package": component, "nested_components": extra_data})
+        recursive_component_collector(component.components, collected)
+    return collected
+
+
+def resolve_license(license):
+    """
+    Return license expression/id/name from license item.
+    """
+    if "expression" in license:
+        return license["expression"]
+    elif "id" in license["license"]:
+        return license["license"]["id"]
+    else:
+        return license["license"]["name"]
+
+
+def get_declared_licenses(licenses):
+    """
+    Return resolved license from list of LicenseChoice.
+    """
+    if not licenses:
+        return ""
+
+    return "\n".join(
+        [resolve_license(license) for license in bom_attributes_to_dict(licenses)]
+    )
```

Comment:
Storing the list in a variable before the join will help with readability, for example:
```
resolved_licenses = [resolve_license(license) for license in bom_attributes_to_dict(licenses)]
return "\n".join(resolved_licenses)
```